File: data-structures-mastery/02_the_array/c/test_fixed_array.c
#include "fixed_array.h"
#include <assert.h>
#include <stdio.h>

int main(void) {
    FixedArray arr;
    array_init(&arr);

    /* a fresh array has length 0 */
    assert(array_length(&arr) == 0);

    /* push fills slots in order and reports success */
    for (int i = 0; i < FIXED_ARRAY_CAPACITY; i++) {
        assert(array_push(&arr, i) == 1);
    }
    assert(array_length(&arr) == FIXED_ARRAY_CAPACITY);

    /* pushing past capacity must fail and must NOT change length */
    assert(array_push(&arr, 999) == 0);
    assert(array_length(&arr) == FIXED_ARRAY_CAPACITY);

    /* get returns the exact values pushed, in order */
    int value;
    for (int i = 0; i < FIXED_ARRAY_CAPACITY; i++) {
        assert(array_get(&arr, (size_t)i, &value) == 1);
        assert(value == i);
    }

    /* get on an out-of-bounds index fails and leaves *out_value untouched */
    value = -1;
    assert(array_get(&arr, 999, &value) == 0);
    assert(value == -1);

    /* set overwrites an in-bounds slot */
    assert(array_set(&arr, 3, 777) == 1);
    array_get(&arr, 3, &value);
    assert(value == 777);

    /* set on an out-of-bounds index fails */
    assert(array_set(&arr, 999, 1) == 0);

    /* stride proof, reused from Module 01: neighboring slots are
     * exactly sizeof(int) bytes apart */
    const int *p0 = array_raw_address(&arr, 0);
    const int *p1 = array_raw_address(&arr, 1);
    assert((const char *)p1 - (const char *)p0 == (ptrdiff_t)sizeof(int));

    printf("All C tests passed.\n");
    return 0;
}

File: data-structures-mastery/02_the_array/python/test_fixed_array.py
from fixed_array import FixedArray, FixedArrayFullError, FIXED_ARRAY_CAPACITY


def test_starts_empty():
    arr = FixedArray()
    assert len(arr) == 0


def test_push_fills_in_order():
    arr = FixedArray()
    for i in range(FIXED_ARRAY_CAPACITY):
        arr.push(i)
    assert len(arr) == FIXED_ARRAY_CAPACITY
    for i in range(FIXED_ARRAY_CAPACITY):
        assert arr.get(i) == i


def test_push_past_capacity_raises_and_does_not_grow():
    arr = FixedArray()
    for i in range(FIXED_ARRAY_CAPACITY):
        arr.push(i)
    try:
        arr.push(999)
        assert False, "expected FixedArrayFullError"
    except FixedArrayFullError:
        pass
    assert len(arr) == FIXED_ARRAY_CAPACITY


def test_get_out_of_bounds_raises():
    arr = FixedArray()
    arr.push(1)
    try:
        arr.get(5)
        assert False, "expected IndexError"
    except IndexError:
        pass


def test_set_overwrites_in_bounds_slot():
    arr = FixedArray()
    for i in range(3):
        arr.push(i)
    arr.set(1, 777)
    assert arr.get(1) == 777


def test_set_out_of_bounds_raises():
    arr = FixedArray()
    arr.push(1)
    try:
        arr.set(5, 1)
        assert False, "expected IndexError"
    except IndexError:
        pass


if __name__ == "__main__":
    test_starts_empty()
    test_push_fills_in_order()
    test_push_past_capacity_raises_and_does_not_grow()
    test_get_out_of_bounds_raises()
    test_set_overwrites_in_bounds_slot()
    test_set_out_of_bounds_raises()
    print("All Python tests passed.")
